src/System.Private.CoreLib/src/System/Threading/ClrThreadPool.HillClimbing.cs

```diff
@@ -0,0 +1,385 @@
+                // sample interval.  The count produced by each thread includes an initial work item that may have 
+                // started well before the start of the interval, and each thread may have been running some new 
+                // work item for some time before the end of the interval, which did not yet get counted.  So
+                // our count is going to be off by +/- threadCount workitems.
+                //
+                // The exception is that the thread that reported to us last time definitely wasn't running any work
+                // at that time, and the thread that's reporting now definitely isn't running a work item now.  So
+                // we really only need to consider threadCount-1 threads.
+                //
+                // Thus the percent error in our count is +/- (threadCount-1)/numCompletions.
+                //
+                // We cannot rely on the frequency-domain analysis we'll be doing later to filter out this error, because
+                // of the way it accumulates over time.  If this sample is off by, say, 33% in the negative direction,
+                // then the next one likely will be too.  The one after that will include the sum of the completions
+                // we missed in the previous samples, and so will be 33% positive.  So every three samples we'll have 
+                // two "low" samples and one "high" sample.  This will appear as periodic variation right in the frequency
+                // range we're targeting, which will not be filtered by the frequency-domain translation.
+                //
+                if (_totalSamples > 0 && ((currentThreadCount - 1.0) / numCompletions) >= _maxSampleError)
+                {
+                    // not accurate enough yet.  Let's accumulate the data so far, and tell the ThreadPool
+                    // to collect a little more.
+                    _accumulatedSampleDuration = sampleDuration;
+                    _accumulatedCompletionCount = numCompletions;
+                    return (currentThreadCount, 10);
+                }
+
+                //
+                // We've got enouugh data for our sample; reset our accumulators for next time.
+                //
+                _accumulatedSampleDuration = 0;
+                _accumulatedCompletionCount = 0;
+
+                //
+                // Add the current thread count and throughput sample to our history
+                //
+                double throughput = numCompletions / sampleDuration;
+                //Worker Thread Adjustment Sample event
+
+                int sampleIndex = _totalSamples % _samplesToMeasure;
+                _samples[sampleIndex] = throughput;
+                _threadCounts[sampleIndex] = currentThreadCount;
+                _totalSamples++;
+
+                //
+                // Set up defaults for our metrics
+                //
+                (double real, double imaginary) threadWaveComponent = (0, 0);
```

Comment:
I think it would be better to have a `Complex` struct that simplifies the math on these values, as it is done in CoreCLR